src/main/java/com/github/manolo8/darkbot/gui/MapDrawer.java

```diff
@@ -79,10 +87,27 @@ public void setup(Main main) {
 
     protected void onPaint() {
         //background image is drawn first
-        if (main.config.BOT_SETTINGS.CUSTOM_BACKGROUND.ENABLED && main.config.BOT_SETTINGS.CUSTOM_BACKGROUND.IMAGE.getImage() != null) {
-            mapGraphics.g2.setComposite(AlphaComposite.getInstance(AlphaComposite.SRC_OVER, (float) main.config.BOT_SETTINGS.CUSTOM_BACKGROUND.OPACITY));
-            mapGraphics.g2.drawImage(main.config.BOT_SETTINGS.CUSTOM_BACKGROUND.IMAGE.getImage(), 0, 0, mapGraphics.width, mapGraphics.height, this);
-            mapGraphics.g2.setComposite(AlphaComposite.getInstance(AlphaComposite.SRC_OVER, 1f));
+        if (main.config.BOT_SETTINGS.CUSTOM_BACKGROUND.ENABLED) {
+            Image img = null;
+
+            if (main.config.BOT_SETTINGS.CUSTOM_BACKGROUND.USE_GAME_BACKGROUND) {
+                GameMap currentMap = main.hero.getMap();
+
+                if (currentMap != lastMap) {
+                    minimapFuture = currentMap.getId() > 0 ? flashResManager.getBackgroundImage(currentMap) : null;
+                    lastMap = currentMap;
```

Comment:
```suggestion
                    if (minimapFuture != null) future.cancel();
                    this.minimapImg = null;
                    lastMap = currentMap;
                    minimapFuture = currentMap.getId() <= 0 ? null :
                            flashResManager.getBackgroundImage(currentMap)
                                    .thenAccept(result -> this.minimapImg = result)
                                    .whenComplete((result, t) -> this.minimapFuture = null);
```
Cancel any previous future, to avoid them potentially overriding the new image (although in practice due to the pipelined nature of the background thread (no concurrency) this would never occur.
Cleanup the current "minimapImg" (old)
Update last map
Resolve the new future, and when it's done, just set minimapImg to the result. Also when completed (either normally or exceptionally) cleanup the future so only the img itself stays.